unit/testing-utils/require_type.cpp

```diff
@@ -78,3 +93,105 @@ code_typet::parametert require_type::require_parameter(
+/// \param type The type to check
+/// \param type_variables An optional set of type variable identifiers which
+/// should be expected as the type parameters of the generic type.
+/// \return The given type, cast to a java_generic_typet
+const java_generic_typet &require_type::require_java_generic_type_variables(
+  const typet &type,
+  const optionalt<std::initializer_list<irep_idt>> &type_variables)
+{
+  REQUIRE(is_java_generic_type(type));
+  const java_generic_typet &generic_type=to_java_generic_type(type);
+  if(type_variables)
+  {
+    const java_generic_typet::generic_type_variablest &generic_type_vars=
+      generic_type.generic_type_variables();
+    REQUIRE(generic_type_vars.size()==type_variables.value().size());
+    REQUIRE(
+      std::equal(
+        type_variables->begin(),
+        type_variables->end(),
+        generic_type_vars.begin(),
+        [](const irep_idt &type_var_name, const java_generic_parametert &param)
+        {
+          REQUIRE(!is_java_generic_inst_parameter((param)));
+          return param.type_variable().get_identifier()==type_var_name;
+        }));
+  }
+
+  return generic_type;
+}
+
+/// Verify a given type is a java_generic_type, optionally checking
+/// that it's associated type variables match a given set of identifiers
+/// \param type The type to check
+/// \param type_variables An optional set of type variable identifiers which
+/// should be expected as the type parameters of the generic type.
+/// \return The given type, cast to a java_generic_typet
+const java_generic_typet
+&require_type::require_java_generic_type_instantiations(
+  const typet &type,
+  const optionalt<std::initializer_list<irep_idt>> &type_instantiations)
+{
+  REQUIRE(is_java_generic_type(type));
+  const java_generic_typet &generic_type=to_java_generic_type(type);
+  if(type_instantiations)
+  {
+    const java_generic_typet::generic_type_variablest &generic_type_vars=
+      generic_type.generic_type_variables();
+    REQUIRE(generic_type_vars.size()==type_instantiations.value().size());
+    REQUIRE(
+      std::equal(
+        type_instantiations->begin(),
+        type_instantiations->end(),
+        generic_type_vars.begin(),
+        [](const irep_idt &type_id, const java_generic_parametert &param)
+        {
+          REQUIRE(is_java_generic_inst_parameter((param)));
+          return param.subtype()==symbol_typet(type_id);
+        }));
+  }
+
+
+  return generic_type;
+}
+
+/// Verify a given type is a java_generic_parameter, optionally checking
+/// that it's associated type variables match a given set of identifiers
+/// \param type The type to check
+/// \param type_variables An optional set of type variable identifiers which
+/// should be expected as the type parameters of the generic type.
+/// \return The given type, cast to a java_generic_typet
+const java_generic_parametert
+&require_type::require_java_generic_parameter_variables(
```

Comment:
It would be great to have a similar function for instantiated parameters, something like `require_java_generic_parameter_instantiated`. Then these two functions can be called within the two functions above, for checking generic type and the parameters.